sqlstats,ui: collect and display count of generic query plans

mgartner opened the pull request:
The UI now displays the number of times a plan was executed as a generic
query plan. This is visible in both the "Explain Plans" table and the
plan details page.

Fixes #144585

Release note (ui): The "Explain Plans" table of the statement details
page and the plan details page now indicate whether or not the plan was
a generic query plan.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144008)
<!-- Reviewable:end -->

mgartner committed:
sqlstats,ui: collect and display count of generic query plans

The UI now displays the number of times a plan was executed as a generic
query plan. This is visible in both the "Explain Plans" table and the
plan details page.

Release note (ui): The "Explain Plans" table of the statement details
page and the plan details page now indicate whether or not the plan was
a generic query plan.

mgartner commented:
@xinhaoz @kyle-a-wong I think this is ready for a look. I have no experience with UI changes like this, so please scrutinize thoroughly. Are there tests I should be adding?

Some screen shots are below. cc @dikshant does this LGTY?

<img width="949" alt="image" src="https://github.com/user-attachments/assets/4d302f4c-26ae-40c2-8bbf-a37a17246090" />

<img width="1017" alt="image" src="https://github.com/user-attachments/assets/a0e00792-b1f8-4f16-ba3b-6dbbbb26a2e3" />

<img width="1267" alt="image" src="https://github.com/user-attachments/assets/194b55c5-e51c-4927-bacf-27f0f9a9e9bf" />

kyle-a-wong reviewed (commented):
Overall this looks good.

I have a question regarding generic plans, is this a new feature in 25.2 or is it just these sql stats that are new? If the latter, it may lead to a bit of confusion on the customer side because the generic count will start at 0 while the total count is > 0 for already persisted stats. Once statements start recording the counts, it may show some stats as `X Yes / Y No` until the older starts are evicted from the table. 

If this is an issue, we could introduce a new cluster setting that disables these fields by default in the UI (while still recording in sql stats) and allow customers to turn it on, or just add the UI part in a future version (25.2.1 or something).

kyle-a-wong commented:
@mgartner Can you add a release note that mentions that this will be added to sql stats?

xinhaoz reviewed (commented):
LGTM! I think we may want some commentary from Ina or Kevin here on the wording / formatting and how to handle missing values if generic query plans have been around for some releases already, as per the concern in Kyle's comment.

I'm fine if we just merge now and obs can fix above things in follow ups as the discussion may be more involved.

mgartner commented:
>Once statements start recording the counts, it may show some stats as X Yes / Y No until the older starts are evicted from the table.

This should be fine. Generic query plans were released in 24.2.0 (and backported to some earlier patch releases) but have been disabled by default until now. If some plans have `X Yes, Y No` then that'd be accurate. Most of the time that generic query plans are used, the query plan is different than the non-generic plan, so it'll have a different plan gist/hash. So most commonly, a user will not see `X Yes / Y No`, they'll see an additional plan in the list.

The only time it will be inaccurate is if users have manually turned on generic query plans, which only a handful have, I assume. So I don't think it is something to worry too much about, but I defer you you all.

> Can you add a release note that mentions that this will be added to sql stats?

Yes, done.

dikshant commented:
> Some screen shots are below. cc @dikshant does this LGTY?

LGTM!

kyle-a-wong commented:
> Most of the time that generic query plans are used, the query plan is different than the non-generic plan, so it'll have a different plan gist/hash. So most commonly, a user will not see `X Yes / Y No`, they'll see an additional plan in the list.

In this case, the "generic plan" hash won't show as "Yes" (it will always be `X Yes / Y No`) in the UI because genericCount == totalCount won't be true until the older statement stats are deleted.

cc @kevin-v-ngo @innabs

mgartner commented:
> In this case, the "generic plan" hash won't show as "Yes" (it will always be X Yes / Y No) in the UI because genericCount == totalCount won't be true until the older statement stats are deleted.

Agreed. But like I said, this is only going to happen for the small minority of customers which enabled generic query plans manually—they are disable by default on all previously released versions.

mgartner commented:
TFTRs!

bors+

mgartner commented:
bors r+
